packages/compartment-mapper/src/policy.js

```diff
@@ -0,0 +1,305 @@
+      );
+    }
+  }
+};
+
+/**
+ * Returns the policy applicable to the canonicalName - either by taking from user
+ * supplied policy or returning localPolicy if user didn't specify one at runtime.
+ *
+ * @param {PolicyIdentityComponents} identityComponents - a key in the policy resources spec
+ * @param {Object|undefined} policy - user supplied policy
+ * @returns {Object|undefined} policy fragment if policy was specified
+ */
+export const getPolicyFor = (identityComponents, policy) => {
+  const canonicalName = generateCanonicalName(identityComponents);
+  if (!policy) {
+    return undefined;
+  }
+  if (canonicalName === ATTENUATORS_COMPARTMENT) {
+    return {
+      packages: detectAttenuators(policy).reduce((packages, specifier) => {
+        packages[specifier] = true;
+        return packages;
+      }, {}),
+    };
+  }
+  if (policy.resources && policy.resources[canonicalName]) {
+    validateDependencies(policy, canonicalName);
+    return policy.resources[canonicalName];
+  } else {
+    console.warn(
+      `No policy for '${canonicalName}', omitting from compartment map.`,
+    );
+    return undefined;
+  }
+};
+
+const getGlobalsList = myPolicy => {
+  if (!myPolicy.globals) {
+    return [];
+  }
+  // TODO: handle 'write' policy
+  return entries(myPolicy.globals)
+    .filter(([_k, v]) => v)
+    .map(([k, _v]) => k);
+};
+
+/**
+ * Filters available globals and returns a copy according to the policy
+ *
+ * @param {Object} globals
+ * @param {Object} localPolicy
+ * @returns {Object} limitedGlobals
+ */
+export const getAllowedGlobals = (globals, localPolicy) => {
+  if (!localPolicy) {
+    return globals;
+  }
+  const list = getGlobalsList(localPolicy);
+  return copyGlobals(globals, list);
+};
+
+/**
+ * Throws if importing of the specifier is not allowed by the policy
+ *
+ * @param {string} specifier
+ * @param {Object} compartmentDescriptor
+ * @param {Object} [info]
+ */
+export const gatekeepModuleAccess = (
+  specifier,
+  compartmentDescriptor,
+  info,
+) => {
+  const { policy, modules } = compartmentDescriptor;
+  if (!policy) {
+    return;
+  }
+
+  if (!info.exit) {
+    if (!modules[specifier]) {
+      throw Error(
+        `Importing '${specifier}' was not allowed by policy packages:${JSON.stringify(
+          policy.packages,
+        )}`,
+      );
+    }
+    return;
+  }
+
+  if (!policy.builtins || !policy.builtins[specifier]) {
+    throw Error(
+      `Importing '${specifier}' was not allowed by policy 'builtins':${JSON.stringify(
+        policy.builtins,
+      )}`,
+    );
+  }
+};
+
+function attenuateModule({ attenuators, name, params, originalModule }) {
+  if (!attenuators) {
+    // TODO: figure out what to pass here or where to move this check
+    throw Error(
+      `Attenuation '${name}' in policy doesn't have a corresponding implementation.`,
+    );
+  }
+  const attenuationCompartment = new Compartment(
```

Comment:
Is it the case that attenuation is incompatible with the speak-no-eval bundling @kumavis is working on?  If so, would that be worth a comment in one of the two PRs?

Reply:
There's no going from source text to execution involved here. If a mock StaticModuleRecord-shaped output is returned from importHook would `eval` be called anywhere?